File: Tests/Categories/TestUIActionSheet.m
#import "UIActionSheet+AnyPromise.h"
@import PromiseKit;
@import XCTest;
@import UIKit;

@implementation Test_UIActionSheet_Objc : XCTestCase {
    UIViewController *rootvc;
}

- (void)setUp {
    rootvc = [UIApplication sharedApplication].keyWindow.rootViewController = [UIViewController new];
}

- (void)tearDown {
    [UIApplication sharedApplication].keyWindow.rootViewController = nil;
}

// fulfills with buttonIndex
- (void)test1 {
    id ex = [self expectationWithDescription:@""];

    UIActionSheet *sheet = [UIActionSheet new];
    [sheet addButtonWithTitle:@"0"];
    [sheet addButtonWithTitle:@"1"];
    sheet.cancelButtonIndex = [sheet addButtonWithTitle:@"2"];
    [sheet promiseInView:rootvc.view].then(^(id obj){
        XCTAssertEqual([obj integerValue], 1);
        [ex fulfill];
    });
    PMKAfter(0.1).then(^{
        [sheet dismissWithClickedButtonIndex:1 animated: NO];
    });
    [self waitForExpectationsWithTimeout:10 handler: nil];
}

// cancel button presses are cancelled errors
- (void)test2 {
    id ex = [self expectationWithDescription:@""];

    UIActionSheet *sheet = [UIActionSheet new];
    [sheet addButtonWithTitle:@"0"];
    [sheet addButtonWithTitle:@"1"];
    sheet.cancelButtonIndex = [sheet addButtonWithTitle:@"2"];
    [sheet promiseInView:rootvc.view].catchWithPolicy(PMKCatchPolicyAllErrors, ^(NSError *err){
        XCTAssertTrue(err.isCancelled);
        [ex fulfill];
    });
    PMKAfter(0.1).then(^{
        [sheet dismissWithClickedButtonIndex:2 animated: NO];
    });
    [self waitForExpectationsWithTimeout:10 handler: nil];
}

// single button UIActionSheets don't get considered cancelled
- (void)test3 {
    id ex = [self expectationWithDescription:@""];

    UIActionSheet *sheet = [UIActionSheet new];
    [sheet addButtonWithTitle:@"0"];
    [sheet promiseInView:rootvc.view].then(^(id obj){
        XCTAssertEqual([obj integerValue], 0);
        [ex fulfill];
    });
    PMKAfter(0.1).then(^{
        [sheet dismissWithClickedButtonIndex:0 animated: NO];
    });
    [self waitForExpectationsWithTimeout:10 handler: nil];
}

// single button UIActionSheets don't get considered cancelled unless the cancelIndex is set
- (void)test4 {
    id ex = [self expectationWithDescription:@""];

    UIActionSheet *sheet = [UIActionSheet new];
    sheet.cancelButtonIndex = [sheet addButtonWithTitle:@"0"];
    [sheet promiseInView:rootvc.view].catchWithPolicy(PMKCatchPolicyAllErrors, ^(NSError *err){
        XCTAssertTrue(err.isCancelled);
        [ex fulfill];
    });
    PMKAfter(0.1).then(^{
        [sheet dismissWithClickedButtonIndex:0 animated: NO];
    });
    [self waitForExpectationsWithTimeout:10 handler: nil];
}

@end

File: Tests/Categories/TestUIActionSheet.swift
import PromiseKit
import UIKit
import XCTest

class Test_UIActionSheet_Swift: UIKitTestCase {
    // fulfills with buttonIndex
    func test1() {
        let ex = expectation(description: "")

        let sheet = UIActionSheet()
        sheet.addButton(withTitle: "0")
        sheet.addButton(withTitle: "1")
        sheet.cancelButtonIndex = sheet.addButton(withTitle: "2")
        sheet.promise(in: rootvc.view).then { x -> Void in
            XCTAssertEqual(x, 1)
            ex.fulfill()
        }
        after(interval: 0.1).then {
            sheet.dismiss(withClickedButtonIndex: 1, animated: false)
        }
        waitForExpectations(timeout: 3)
    }

    // cancel button presses are cancelled errors
    func test2() {
        let ex = expectation(description: "")

        let sheet = UIActionSheet()
        sheet.addButton(withTitle: "0")
        sheet.addButton(withTitle: "1")
        sheet.cancelButtonIndex = sheet.addButton(withTitle: "2")
        sheet.promise(in: rootvc.view).catch(policy: .allErrors) { err in
            guard let err = err as? UIActionSheet.Error else { return XCTFail() }
            XCTAssertTrue(err == UIActionSheet.Error.cancelled)
            XCTAssertTrue(err.isCancelled)
            ex.fulfill()
        }
        after(interval: 0.1).then {
            sheet.dismiss(withClickedButtonIndex: 2, animated: false)
        }
        waitForExpectations(timeout: 3)
    }

    // single button UIActionSheets don't get considered cancelled
    func test3() {
        let ex = expectation(description: "")

        let sheet = UIActionSheet()
        sheet.addButton(withTitle: "0")
        sheet.promise(in: rootvc.view).then { _ in
            ex.fulfill()
        }
        after(interval: 0.1).then {
            sheet.dismiss(withClickedButtonIndex: 0, animated: false)
        }
        waitForExpectations(timeout: 3)
    }

    // single button UIActionSheets don't get considered cancelled unless the cancelIndex is set
    func test4() {
        let ex = expectation(description: "")

        let sheet = UIActionSheet()
        sheet.cancelButtonIndex = sheet.addButton(withTitle: "0")
        sheet.promise(in: rootvc.view).catch(policy: .allErrors) { _ in
            ex.fulfill()
        }
        after(interval: 0.1).then {
            sheet.dismiss(withClickedButtonIndex: 0, animated: false)
        }
        waitForExpectations(timeout: 3)
    }
}
